apiserver/common/leadership.go

```diff
@@ -103,47 +99,93 @@ func (a *leadershipPinningAPI) PinnedLeadership() (params.PinnedLeadershipResult
 	return result, nil
 }
 
-// TODO (manadart 2018-10-29): Rename the two methods below (and on the client
-// side) to be [Un]PinApplicationLeaders, and derive the list of applications
-// based on the authenticating entity.
-
-// PinMachineApplications pins leadership for applications represented by units
-// running on the auth'd machine.
-func (a *leadershipPinningAPI) PinMachineApplications() (params.PinApplicationsResults, error) {
+// PinApplicationLeaders pins leadership for applications based on the auth
+// tag provided.
+func (a *LeadershipPinning) PinApplicationLeaders() (params.PinApplicationsResults, error) {
 	if !a.authorizer.AuthMachineAgent() {
 		return params.PinApplicationsResults{}, ErrPerm
 	}
-	return a.pinMachineAppsOps(a.pinner.PinLeadership)
-}
 
-// UnpinMachineApplications unpins leadership for applications represented by
-// units running on the auth'd machine.
-func (a *leadershipPinningAPI) UnpinMachineApplications() (params.PinApplicationsResults, error) {
-	if !a.authorizer.AuthMachineAgent() {
+	tag := a.authorizer.GetAuthTag()
+	switch tag.Kind() {
+	case names.MachineTagKind:
+		return a.pinMachineApplications(tag)
+	default:
 		return params.PinApplicationsResults{}, ErrPerm
 	}
-	return a.pinMachineAppsOps(a.pinner.UnpinLeadership)
 }
 
-// pinMachineAppsOps runs the input pin/unpin operation against all
-// applications represented by units on the authorised machine.
-// An assumption is made that the validity of the auth tag has been verified
-// by the caller.
-func (a *leadershipPinningAPI) pinMachineAppsOps(op func(string, string) error) (params.PinApplicationsResults, error) {
-	result := params.PinApplicationsResults{}
+// UnpinApplicationLeaders unpins leadership for applications based on the auth
+// tag provided.
+func (a *LeadershipPinning) UnpinApplicationLeaders() (params.PinApplicationsResults, error) {
+	if !a.authorizer.AuthMachineAgent() {
+		return params.PinApplicationsResults{}, ErrPerm
+	}
 
 	tag := a.authorizer.GetAuthTag()
-	m, err := a.st.Machine(tag.Id())
+	switch tag.Kind() {
+	case names.MachineTagKind:
+		return a.unpinMachineApplications(tag)
+	default:
+		return params.PinApplicationsResults{}, ErrPerm
+	}
+}
+
+// GetMachineApplicationNames returns the applications associated with a
+// machine.
+func (a *LeadershipPinning) GetMachineApplicationNames(id string) ([]string, error) {
+	m, err := a.st.Machine(id)
 	if err != nil {
-		return result, errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
 	apps, err := m.ApplicationNames()
 	if err != nil {
-		return result, errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
+	return apps, nil
+}
+
+// PinMachineApplicationsByName takes a slice of application names and attempts
+// to pin them accordingly.
+func (a *LeadershipPinning) PinMachineApplicationsByName(tag names.Tag, appNames []string) (params.PinApplicationsResults, error) {
```

Comment:
These too are no longer specific to a machine entity, so the name should change.
For all the `tag` arguments, I think we should rename them to something like `requiredBy`.